### DISCUSSION TITLE
onBeforeRender in vike-react takes long time to do especially when querying data

### DISCUSSION THREAD

**Discussion by BanDroid** [vikejs/vike#1066] [Discussion]
please optimize onBeforeRender, i'm running a simple mongodb query in there, but takes really long time to load the page, i dont know why

something like this

```javascript
const query = await MyModel.findById(id);
return {
  pageContext: {
    pageProps: {},
    title: query.title
  }
}
```

it is simple query but takes long time to load the page.

and also, onBeforeRender does not show loading indicator in client browser when navigating between pages, is this default behavior in vike-react to use client routing?

**Comment by brillout** [comment]
That's very likely a user land issue.

> really long time to load the page

How long? Vite-plugin-ssr should be logging a warning if your hook is taking more than 4 seconds.

> and also, onBeforeRender does not show loading indicator in client browser when navigating between pages, is this default behavior in vike-react to use client routing?

I agree that'd be nice. Even better would be to include this to Bati and/or Stem (because Vike React should be ejected only as last resort).
